### PR TITLE
Fix promtail deadlock

### LINKED ISSUE #5162: Promtail stuck on mutex while tailing files.
Since #4794 we have seen many cases of promtail getting stuck and not tailing file anymore.

https://github.com/grafana/loki/pull/5043 was supposed to fix it but we've seen case today again.

The goroutine stack trace seems to shows that it's a deadlock.

[code-stdin-zag.txt](https://github.com/grafana/loki/files/7881772/code-stdin-zag.txt)

[pprof.promtail.goroutine.003.pb.gz](https://github.com/grafana/loki/files/7881774/pprof.promtail.goroutine.003.pb.gz)

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

This PR fixes a deadlock in Promtail which occurred when targets were removed from the target groups by the discovery manager.

The problem was that upon removal of a target it was removed from the targets list, but the correlating file event watcher channel was not removed from from the watchers list. This caused events to be sent to channels that were not read any more and therefore resulting in a deadlock.

**Which issue(s) this PR fixes**:
Fixes #5162

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: clients/pkg/promtail/targets/file/filetargetmanager.go [hunk 1]
```diff
@@ -319,6 +319,10 @@ func (s *targetSyncer) sync(groups []*targetgroup.Group, targetEventHandler chan
 			target.Stop()
 			s.metrics.targetsActive.Add(-1.)
 			delete(s.targets, key)
+
+			// close related file event watcher
```
Comment (cyriltovena) [inline clients/pkg/promtail/targets/file/filetargetmanager.go:323]: cc @slim-bean :)

**Review by cyriltovena — APPROVED** [review]
LGTM

File: clients/pkg/promtail/targets/file/filetargetmanager.go [hunk 2]
```diff
@@ -319,6 +319,10 @@ func (s *targetSyncer) sync(groups []*targetgroup.Group, targetEventHandler chan
...
+			close(s.fileEventWatchers[target.path])
```
[the remark this answers is not in the document]
Comment (chaudum) [reply clients/pkg/promtail/targets/file/filetargetmanager.go:324]: Probably a good idea, on the other hand, there went something wrong beforehand if the member does not exist.

[the remark this answers is not in the document]
Comment (cyriltovena) [reply clients/pkg/promtail/targets/file/filetargetmanager.go:324]: yeah good idea !

[the remark this answers is not in the document]
Comment (cyriltovena) [reply clients/pkg/promtail/targets/file/filetargetmanager.go:324]: It's the only place we don't iterate. Let's do it.

**Review by kavirajk — APPROVED** [review]
LGTM

**Pull request merged**